content/blog/2016/2016-12-13-jenkins-puppet-enterprise-plugin.adoc

```diff
@@ -0,0 +1,271 @@
+generates and collects across every system Puppet is installed on. PuppetDB's
+provides a sweet query language called
+link:https://docs.puppet.com/puppetdb/4.3/api/query/v4/pql.html[PQL.] With PQL,
+you can ask complex questions of your infrastructure such as "How production
+Red Hat systems are there with the openssl package installed?" or "What are are
+the us-west-2c nodes with the MyApp role that were created in the last 24
+hours?"
+
+This can be a pretty powerful tool for parts of your pipeline where you need to
+perform specific operations on subsets of the infrastructure like draining a
+loadbalancer.
+
+To perform a query, use the puppet.query method.
+
+----
+results = puppet.query 'inventory[certname] { facts.os.name = "RedHat" and facts.ec2_metadata.placement.availability-zone = "us-west-2c" and uptime_hours < 24 }'
+----
+
+The value returned is an array of items matching the query. The results can be
+iterated on, and even passed to a series of puppet.job calls. For example, the
+following code will query all nodes in production that experienced a failure on
+the last Puppet run, break the resulting list into groups of 3, and runs Puppet
+on each sub group.
+
+----
+results = puppet.query 'nodes { latest_report_status = "failed" and catalog_environment = "production"}'
+nodes = []
+for (Map node : results) {
+  nodes.add(node.certname)
+}
+nodesubgroups = nodes.collate(3) //Break results into groups of 3
+for (String certnames : nodesubgroups ) {
+  puppet.job 'production', nodes: certnames
+}
+----
+
+Note that once you can use closures in Pipeline scripts, doing the above
+example will be much simpler.
+
+== Creating an orchestrator job
+
+The orchestration service in Puppet Enterprise is a powerful tool that enables
+orchestrated Puppet runs across as broad or as targeted an infrastructure as
+you need at different parts of a pipeline. You can use the orchestrator to
+update applications in an environment, or update a specific list of nodes, or
+update nodes across a set of nodes that match certain criteria. In each
+scenario, Puppet will always push distributed changes in the correct order by
+respecting the cross-node dependencies.
+
+To create a job in the Puppet orchestrator from a Jenkins pipeline, use the
+puppet.job method. The puppet.job method will create a new orchestrator job,
+monitor the job for completion, and determine if any Puppet runs failed. If
+there were failures, the pipeline will fail.
+
+To run Puppet against all of production:
+
+----
+puppet.job 'production'
+----
+
+To run Puppet against instances of an application in production:
+
+----
+puppet.job 'production', application: 'Myapp'
+----
+
+To run Puppet against nodes db.example.com, appserver01.example.com, and appserver02.example.com:
+
+----
+puppet.job 'production', nodes: ['db.example.com','appserver01.example.com','appserver02.example.com']
+----
+
+To run Puppet against all Red Hat nodes in the AWS us-west-2c region that were
```

Comment:
That was kind of the point. It's using repetitive language to give a sense of endless ways to target the infrastructure you want to change.  But I can see your point that it can be tiresome to read.  I'll reword this section.